src/python/pants/engine/internals/visibility.py

```diff
@@ -0,0 +1,58 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    FieldSet,
+    ValidatedDependencies,
+    ValidateDependenciesRequest,
+    VisibilityField,
+    VisibilityViolationError,
+    WrappedTarget,
+    WrappedTargetRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.util.strutil import softwrap
+
+
+class ValidateVisibilityRulesFieldSet(FieldSet):
+    required_fields = (VisibilityField,)
+
+
+class ValidateVisibilityRulesRequest(ValidateDependenciesRequest):
+    field_set_type = ValidateVisibilityRulesFieldSet
```

Comment:
We could have a check that enforces all Target types to have a visibility field (and all that include the COMMON_TARGET_FIELDS get it).

I think this would be a potential gotcha for plugin authors, which a note here would be of moot help.

Reply:
This is addressed by #15876 so that we can register the visibility field as a plugin field directly on `Target` and it will apply to _all_ targets.